From 9784ad6b0a2dfd478bf9299d0a0f36450ce7f239 Mon Sep 17 00:00:00 2001
From: Amir Vadai <avadai@redhat.com>
Date: Thu, 11 Sep 2014 12:01:48 -0400
Subject: [ethernet] mlx4: Reduce memory consumption on kdump kernel

Message-id: <cf0b4378087d6f9c67447f7fcc18b027498f9f1e.1410425016.git.avadai@redhat.com>
Patchwork-id: 93685
O-Subject: [Patch RHEL7.1 V1 172/197] net/mlx4_en: Reduce memory consumption on kdump kernel
Bugzilla: 1107617 1107618 1107619 1107620
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

From: Amir Vadai <amirv@mellanox.com>

When memory is limited, reduce number of rx and tx rings.

Signed-off-by: Amir Vadai <amirv@mellanox.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit ea1c1af1396cac9f8a1160acdac17f80e4b4f2c4)

Bugzilla: 1121698
Signed-off-by: Amir Vadai <avadai@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/mellanox/mlx4/en_main.c b/drivers/net/ethernet/mellanox/mlx4/en_main.c
index f953c1d..3626fdf 100644
--- a/drivers/net/ethernet/mellanox/mlx4/en_main.c
+++ b/drivers/net/ethernet/mellanox/mlx4/en_main.c
@@ -129,8 +129,10 @@ static int mlx4_en_get_profile(struct mlx4_en_dev *mdev)
  int i;
 
  params->udp_rss = udp_rss;
- params->num_tx_rings_p_up = min_t(int, num_online_cpus(),
-   MLX4_EN_MAX_TX_RING_P_UP);
+ params->num_tx_rings_p_up = mlx4_low_memory_profile() ?
+  MLX4_EN_MIN_TX_RING_P_UP :
+  min_t(int, num_online_cpus(), MLX4_EN_MAX_TX_RING_P_UP);
+
  if (params->udp_rss && !(mdev->dev->caps.flags
      & MLX4_DEV_CAP_FLAG_UDP_RSS)) {
   mlx4_warn(mdev, "UDP RSS is not supported on this device\n");
diff --git a/drivers/net/ethernet/mellanox/mlx4/en_rx.c b/drivers/net/ethernet/mellanox/mlx4/en_rx.c
index 3b62b0b..1f394c4 100644
--- a/drivers/net/ethernet/mellanox/mlx4/en_rx.c
+++ b/drivers/net/ethernet/mellanox/mlx4/en_rx.c
@@ -334,8 +334,9 @@ void mlx4_en_set_num_rx_rings(struct mlx4_en_dev *mdev)
         dev->caps.comp_pool/
         dev->caps.num_ports) - 1;
 
-  num_rx_rings = min_t(int, num_of_eqs,
-         netif_get_num_default_rss_queues());
+  num_rx_rings = mlx4_low_memory_profile() ? MIN_RX_RINGS :
+   min_t(int, num_of_eqs,
+         netif_get_num_default_rss_queues());
   mdev->profile.prof[i].rx_ring_num =
    rounddown_pow_of_two(num_rx_rings);
  }
diff --git a/drivers/net/ethernet/mellanox/mlx4/mlx4_en.h b/drivers/net/ethernet/mellanox/mlx4/mlx4_en.h
index 065fdce..f7b149b 100644
--- a/drivers/net/ethernet/mellanox/mlx4/mlx4_en.h
+++ b/drivers/net/ethernet/mellanox/mlx4/mlx4_en.h
@@ -121,6 +121,7 @@ enum {
 #define MLX4_EN_MIN_TX_SIZE (4096 / TXBB_SIZE)
 
 #define MLX4_EN_SMALL_PKT_SIZE  64
+#define MLX4_EN_MIN_TX_RING_P_UP 1
 #define MLX4_EN_MAX_TX_RING_P_UP 32
 #define MLX4_EN_NUM_UP   8
 #define MLX4_EN_DEF_TX_RING_SIZE 512
-- 
1.7.1